!!! TASK: Add workspace content stream mapping to content graph projection

bwaidelich opened the pull request:
Previously the `DoctrineDbalContentGraphProjection` accessed the `workspace` table of a different projection in order to resolve the workspace<->content stream mapping.

This change adds the `workspace` table to the content graph projection and uses that instead for the resolution.

**Note:** This is not a breaking change because it comes with a migration and does not require a replay, but a
```shell
./flow cr:setup
```
is needed in order to apply that!

Related: #5038
Replaces: https://github.com/neos/neos-development-collection/pull/5041


Todos

- [x] discuss REMOVED = False flag when fetching content streams [link](https://github.com/neos/neos-development-collection/pull/5096/commits/80a302fedfe6bbeea4c62ea664a27598f29ab8c6)
- [x] ~discuss  `@deprecated This event will never be emitted, and it is ignored in the core projections. This implementation is just kept for backwards-compatibility` because its not true, the event `WorkspaceWasRenamed` will still be emitted.~ note bw: "emitted" means that the event is published. This change won't remove those events from the existing event streams, but no new events should occur
- [x] discuss `EmbeddsContentStream` vs extraction [link](https://github.com/neos/neos-development-collection/pull/5096/commits/d0839e4fcdc13db5cfe410a9fa9d7d3e67a3b23b)
- [x] discuss readded runtime cache [link](https://github.com/neos/neos-development-collection/pull/5096/commits/2a781d24aefe0218df66688b00ea60627ece739c)
   - [x] discuss if `getWorkspaces` should fill the `findWorkspaceByName` cache 
   - [x] -> extract in separate pr
- [x] discuss b/c layer, is keeping the method `getWorkspaceFinder` enough? [link](https://github.com/neos/neos-development-collection/pull/5096/commits/a3ab868cad3b5c7e4a97d2f2394435ceb0947915)
- [x] discuss `getWorkspaces` vs `findWorkspaces` naming
- [x] discuss dropping old dangeling tables?
   - [x] -> not a problem for beta phase maybe document it somewhere
- [x] followup move internal classes to not be first level and catch the users eye :) [link](https://github.com/neos/neos-development-collection/pull/5096#discussion_r1646601364)
  - [x] -> will be done via https://github.com/neos/neos-development-collection/pull/5272

bwaidelich committed:
TASK: Add workspace content stream mapping to content graph projection

Previously the `DoctrineDbalContentGraphProjection` accessed the `workspace` table of a different projection in order to resolve the workspace<->content stream mapping.

This change adds the `workspace` table to the content graph projection and uses that instead for the resolution.

**Note:** This is not a breaking change because it comes with a migration and does not require a replay, but a
    ./flow cr:setup
is needed in order to apply that!

Related: #5038

bwaidelich commented:
Draft, because it requires some more love

bwaidelich committed:
!!! TASK: Replace dedicated Workspace and ContentStream projections

wip

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -638,6 +687,71 @@ private function whenSubtreeWasUntagged(SubtreeWasUntagged $event): void
+        $this->updateBaseWorkspace($event->workspaceName, $event->baseWorkspaceName, $event->newContentStreamId);
+    }
+
+    private function whenWorkspaceRebaseFailed(WorkspaceRebaseFailed $event): void
+    {
+        $this->markWorkspaceAsOutdatedConflict($event->workspaceName);
+    }
```
do we even need that still as the event is deprecated?

https://github.com/neos/neos-development-collection/pull/4965

will we carry that on forever :O

bwaidelich replied:
I wasn't sure about this one.. The corresponding events would still appear in existing installations so we should probably deal with it

mhsdesign replied:
yeah fine with me does no harm i guess ... but ignoring them would be fine i think as well... as there is no more `AsOutdatedConflict` only outdated?

mhsdesign replied:
will be resolved via https://github.com/neos/neos-development-collection/pull/5143

mhsdesign reviewed (dismissed):
by reading ... but not thaaaat deep ... maybe +0.5

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/Workspace.php:
```diff
@@ -0,0 +1,112 @@
+
+    private function markWorkspaceAsOutdated(WorkspaceName $workspaceName): void
+    {
+        $this->dbal->executeUpdate('
```
this method is deprecated? They say use `executeStatement` instead?

bwaidelich replied:
check!

bwaidelich committed:
wip

bwaidelich committed:
wip

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepositoryReadModel.php:
```diff
@@ -0,0 +1,142 @@
+
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Core;
```
Im not sure about the namespace, as its first level.

There was a time where only the `ContentRepository` was in the first level and IMO that makes sense since its api and should be used. All the other internal stuff also the `CommandHandlingDependencies` should reside in another namespace i think.

`CommandHandlingDependencies` -> `Neos\ContentRepository\Core\CommandHandler`

but the question is where to put this `ContentRepositoryReadModel`?
Maybe in `Projection`?

bwaidelich replied:
I agree. This should be moved!

mhsdesign replied:
Can also be adjusted when finalising everything in https://github.com/neos/neos-development-collection/pull/5272

mhsdesign committed:
TASK: Post rebase phpstan adjustments

mhsdesign commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteTrait.php:
```diff
@@ -133,20 +134,32 @@ protected function readPayloadTable(TableNode $payloadTable): array
      */
     public function iExpectTheContentStreamToNotExist(string $rawContentStreamId): void
```
this thing here which was introduced with https://github.com/neos/neos-development-collection/commit/433f3698d0c66499666f9a31fdd60d63e6f38c04 doesnt tell the truth. It should be `assertFalse` in the "old" state but now it actually checks that the contentstream exists.

The underlying problem is in the removed state. in this case we want to check that the contentstream does not exist with `removed FALSE` 

>  when pruning content streams, NO_LONGER_IN_USE content streams will be properly cleaned from the graph projection.

and also the latter part is actually not ensured by just looking at the contentstreams ... i think we/i was a bit uncarefull there.

mhsdesign replied:
The cr api will now only expose content streams which are not removed. 
See also slack: https://neos-project.slack.com/archives/C04PYL8H3/p1716906008795729?thread_ts=1716905442.715929&cid=C04PYL8H3

mhsdesign committed:
TASK: Reintroduce content graph / cr read model cache

With https://github.com/neos/neos-development-collection/pull/5246 the cache was naively removed ^^
But with the additional complexity of fetching workspaces etc it makes sense to reintroduce this cache layer.

mhsdesign committed:
TASK: Fix `I expect the content stream "user-cs-identifier" to not exist`

The step was previously introduced via https://github.com/neos/neos-development-collection/commit/433f3698d0c66499666f9a31fdd60d63e6f38c04 but wrongly implemented.

see also https://github.com/neos/neos-development-collection/pull/5096#discussion_r1783514808

mhsdesign committed:
TASK: Use `EmbedsContentStreamId` like previously to `updateContentStreamVersion`

`EmbedsContentStreamId` is available since https://github.com/neos/neos-development-collection/pull/5226

mhsdesign committed:
TASK: Remove obsolete import

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/add-workspace-contentstream-mapping-to-contentgraph

mhsdesign committed:
TASK: Fixup after merge of asset usage refactoring

Initially the exporter got passed the `WorkspaceFinder` and had to fetch the live workspace themselves to determine the content stream id.

Temporarily we then passed the $targetContentStreamId, but some places now need the workspace name as well. Thus we pass the whole $targetWorkspace.

Also a utility `getDependantWorkspaces` was introduced to simplify the `filter` usage from the caller.

mhsdesign committed:
TASK: Adjust to `find*` vs `get*` convention

> "all operations which are slow should be named find*", vs. "all operations which operate on already-loaded data should be named get*".

see https://github.com/neos/neos-development-collection/pull/5279#discussion_r1795281709

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/add-workspace-contentstream-mapping-to-contentgraph

mhsdesign committed:
TASK: Adjust to `find*` vs `get*` convention (in `findContentStreams`)

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/add-workspace-contentstream-mapping-to-contentgraph

bwaidelich committed:
Replace `WorkspaceFinder` usage in `UsageController`

mhsdesign committed:
TASK: Php-stanize UsageController and fix php error

mhsdesign committed:
Revert "TASK: Reintroduce content graph / cr read model cache"

In commit 2a781d24aefe0218df66688b00ea60627ece739c the cache invalidation was partly repaired
But as we removed the content graph cache layer currently in 9.0 https://github.com/neos/neos-development-collection/pull/5246

A reintroduction will be discussed as part of another change.

Related https://github.com/neos/neos-development-collection/issues/5039

bwaidelich committed:
Fix parameter name in content graph projection

mhsdesign commented:
Looks good and works well so far. I also run the Neos Ui e2e and 🔝 🥳 
The migration (in `cr:setup` should probably be tested more extensively .. we especially need to make sure when copying the tables that they have the same schema!!!)

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -0,0 +1,209 @@
+ * @internal only used inside the
+ * @see ContentRepositoryReadModel
+ */
+final readonly class ContentRepositoryReadModelAdapter implements ContentRepositoryReadModelAdapterInterface
```
💡 fyi (for other reviewers) while these namings are not perfect - they are not the final ones either :)
in https://github.com/neos/neos-development-collection/pull/5272 we can discuss the final final final draft:)

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -0,0 +1,209 @@
+                {$this->tableNames->contentStream()}
+            WHERE
+                id = :contentStreamId
+                AND removed = FALSE
```
As discussed with basti, we are sure that the additional `FALSE` filter was missing here and should have also been part of the previous findVersionForContentStream or hasContentStream

mhsdesign replied:
⚠️ Or is there an edge-case with `$commandHandlingDependencies->contentStreamExists` where it IS important that the content stream does not exist if pruned or not?
Like if someone prunes a content stream and then voluntarily creates a new workspaces with it? That would crash? Maybe we can add a low level hacky method `hasContentStream`?

mhsdesign replied:
I rethought this again and again and i think the `removed` state is not just an implementation detail but required to be implemented by the adapter. That means we can and should add it to the `ContentStream` readmodel so it is clearly documented when this state should be encountered.
Further adding it to the readmodel allows us to remove the `findUnusedAndRemovedContentStreamIds` in the adapters and interfaces which would be impossible to document due to their complexity. It really feels good to have that now in the ContentStreamPrunner in the core.

see 041f23926f41fa0ad2dad6e62f16efb45aa2e836

mhsdesign replied:
After discussing this again we found a way to get rid of the removed flag from the readmodel: by diffing the actual found content streams of the event store and the ones that are in the projection. And as a second step im currently fully overhauling the content stream pruner so that we dont have to track neither the removed nor the content stream state at all in the projection and always extract the information from the events: https://github.com/neos/neos-development-collection/pull/5297

So the introduction of `ContentStream::removed` was just a temporary state and will be removed again :)

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -92,7 +109,26 @@ public function __construct(
-        foreach ($this->determineRequiredSqlStatements() as $statement) {
+        $statements = $this->determineRequiredSqlStatements();
+
+        // MIGRATION from 2024-05-25: copy data from "cr_<crid>_p_workspace"/"cr_<crid>_p_contentstream" to "cr_<crid>_p_graph_workspace"/"cr_<crid>_p_graph_contentstream" tables
```
❔ do we need to handle the rare cases of the previous workspace/ content stream projection being out of date and behind?
Id say no and for the case this happened one must issue a full replay.

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -191,16 +240,33 @@ public function apply(EventInterface $event, EventEnvelope $eventEnvelope): void
+            WorkspaceWasRemoved::class => $this->whenWorkspaceWasRemoved($event),
             default => throw new \InvalidArgumentException(sprintf('Unsupported event %s', get_debug_type($event))),
         };
+        if ($event instanceof EmbedsContentStreamId && ContentStreamEventStreamName::isContentStreamStreamName($eventEnvelope->streamName)) {
```
⚠️ this does not look fully safe! Above we use `|| $event instanceof EmbedsContentStreamId;` so the `default` arm must not throw?

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -258,20 +251,25 @@ public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
+    public function getContentGraph(WorkspaceName $workspaceName): ContentGraphInterface
+    {
+        return $this->getContentRepositoryReadModel()->getContentGraphByWorkspaceName($workspaceName);
     }
```
the order of methods was previously probably better

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentStream/ContentStreamFinder.php:
```diff
@@ -1,192 +0,0 @@
-/**
- * Internal - implementation detail of {@see ContentStreamPruner}
- *
- * This projection tracks whether content streams are currently in use or not. Each content stream
- * is first CREATED or FORKED, and then moves to the IN_USE or REBASE_ERROR states; or is removed directly
- * in case of temporary content streams.
- *
- * FORKING: Content streams are forked from a base content stream. It can happen that the base content
- * stream is NO_LONGER_IN_USE, but the child content stream is still IN_USE_BY_WORKSPACE. In this case,
- * the base content stream can go to removed=1 (removed from all projections), but needs to be retained
- * in the event store: If we do a full replay, we need the events of the base content stream before the
- * fork happened to rebuild the child content stream.
- * This logic is done in {@see findUnusedAndRemovedContentStreams}.
- *
- * TEMPORARY content streams: Projections should take care to dispose their temporary content streams,
- * by triggering a ContentStreamWasRemoved event after the content stream is no longer used.
```
maybe we can reuse this lovely comment somewhere:)

mhsdesign commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -4,27 +4,27 @@
 
 /**
- * For internal implementation details, see {@see ContentStreamFinder}.
+ * For internal implementation details, see {@see ContentRepositoryReadModel}.
```
the comment reference is wrong i fear

mhsdesign commented on Neos.ContentRepository.Core/Classes/SharedModel/Workspace/ContentStreams.php:
```diff
@@ -0,0 +1,99 @@
+ * @implements \IteratorAggregate<ContentStream>
+ *
+ * @api
+ */
+
```
```suggestion
 */
```

mhsdesign commented on Neos.ContentRepository.Core/Classes/SharedModel/Workspace/Workspace.php:
```diff
@@ -0,0 +1,63 @@
+        $this->status = $status;
+    }
+
+    public function isRootWorkspace(): bool
```
todo use this method also instead of `$this->baseWorkspaceName !== null`

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Tests/Behavior/Bootstrap/FeatureContext.php:
```diff
@@ -123,26 +126,26 @@ public function iRunTheEventMigration(string $contentStream = null): void
-
-        $eventNormalizer = $this->getObject(EventNormalizer::class);
+
+        // HACK to access the property converter
```
💡  fyi ill clean this up in a followup sometime

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/RoutingTrait.php:
```diff
@@ -185,7 +186,9 @@ public function theMatchedNodeShouldBeInContentStreamAndOriginDimension(string $
         Assert::assertTrue($matchedNodeAddress->workspaceName->isLive());
         Assert::assertSame($nodeAggregateId, $matchedNodeAddress->aggregateId->value);
         // todo useless check?
```
💡 this can and should be migrated at some point to use workspace names in the routing tests

mhsdesign reviewed (commented):
I went through the whole code again and despite a few nitpicks and maybe on crucial thing left to discuss it looks really promising so far (and as said before works also)
So ill push a followup discuss the last things and then its merge ready id say!

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -92,7 +109,26 @@ public function __construct(
+            $this->dbal->getSchemaManager()->tablesExist([$legacyWorkspaceTableName])
+            && !$this->dbal->getSchemaManager()->tablesExist([$this->tableNames->workspace()])
+        ) {
+            $statements[] = 'INSERT INTO ' . $this->tableNames->workspace() . ' (name, baseWorkspaceName, currentContentStreamId, status) SELECT workspaceName AS name, baseWorkspaceName, currentContentStreamId, status FROM ' . $legacyWorkspaceTableName;
```
❔  does the casing not matter? In the legacy projection it was `currentcontentstreamid` ... just wondering about the select

kitsunet replied:
this is a good question:

> Partition, subpartition, column, index, stored routine, event, and resource group names are not case-sensitive on any platform, nor are column aliases

So this should be fine, however table names _could_ be depending on platform so there we need to make sure it matches everywhere.

mhsdesign replied:
still i adjusted it, and also double checked that we dont use the old column names anymore (where we renamed them)

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:
```diff
@@ -107,6 +109,30 @@ private function createReferenceRelationTable(): Table
             ->setPrimaryKey(['name', 'position', 'nodeanchorpoint']);
     }
 
+    private function createWorkspaceTable(): Table
```
💡 i compared the schemas again against the current ones and can confirm besides a few renames of columns its the same. Thus the migration in setup can definitely work.

mhsdesign committed:
TASK: Migrate findUnusedAndRemovedContentStreams from SQL to PHP

Instead of messing with the rather lean API just for this usecase the calculation is done in PHP.

This will be slower and consume more memory, but it should not be an issue even with thousands of content streams for this pure cleanup CLI command.

Further this calculation is better placed in the Core package than to be done in each adapter because writing the specification for the interface would already be too complex.

Additionally, the docs were updated in the `ContentStream` readmodel and now also reference the migrated core `findUnusedAndRemovedContentStreams` logic which also explains the `removed` flag.

mhsdesign committed:
TASK: Use `Workspace->isRootWorkspace()`

mhsdesign committed:
TASK: Prevent `default` arm to throw as we can handle possibly all events with `EmbedsContentStreamId`

mhsdesign committed:
TASK: Move `getContentGraph` again upwards

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/add-workspace-contentstream-mapping-to-contentgraph

mhsdesign committed:
BUGFIX: Fix logic of `Workspace::isRootWorkspace`

LOL

kitsunet commented on Neos.Neos/Classes/Service/LinkingService.php:
```diff
@@ -285,12 +285,12 @@ public function createNodeUri(
         $uriBuilder = clone $controllerContext->getUriBuilder();
         $uriBuilder->setRequest($mainRequest);
-        $createLiveUri = $workspace && $workspace->isPublicWorkspace() && !$resolvedNode->tags->contain(SubtreeTag::disabled());
+        $createLiveUri = $workspace && $nodeAddress->workspaceName->isLive() && !$resolvedNode->tags->contain(SubtreeTag::disabled());
```
This condition seems a bit disjointed now and smells like a conceptual problem. Aka why do I ask the workspaceName if we are live (I know why, but is that sensible?). Anyways nothing we can solve right now, just wanted to point out that it looks a bit strange and the condition read clearer before.

mhsdesign reviewed (approved):
With the adjustments and the planned followup https://github.com/neos/neos-development-collection/pull/5272 i can definitely approve this!!! Thank you for your initial effort in May where we discussed this - its really fun to see such old pr get merged and its ideas still being valid!!!

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:
```diff
@@ -107,6 +109,30 @@ private function createReferenceRelationTable(): Table
+        $workspaceTable = self::createTable($this->tableNames->workspace(), [
+            DbalSchemaFactory::columnForWorkspaceName('name')->setNotnull(true),
+            DbalSchemaFactory::columnForWorkspaceName('baseWorkspaceName')->setNotnull(false),
+            DbalSchemaFactory::columnForContentStreamId('currentContentStreamId')->setNotNull(true),
```
something I noticed while working on another bug is that currently nothing prevents two workspaces having the same contentstream, I think that is dangerous, so we could consider a unique index on this field?

mhsdesign replied:
Same here, maybe as a followup so we can discuss it extensively?

mhsdesign replied:
-> https://github.com/neos/neos-development-collection/pull/5296

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:
```diff
@@ -107,6 +109,30 @@ private function createReferenceRelationTable(): Table
+    private function createContentStreamTable(): Table
+    {
+        return self::createTable($this->tableNames->contentStream(), [
+            DbalSchemaFactory::columnForContentStreamId('id')->setNotnull(true),
```
primary key?

mhsdesign replied:
Good idea, though in a followup id say, this change is big enough already and this was not part of the `ContentStreamProjection`.

mhsdesign replied:
-> https://github.com/neos/neos-development-collection/pull/5296

mhsdesign committed:
TASK: Adjust casings of columns

Just for cosmetics: https://github.com/neos/neos-development-collection/pull/5096#discussion_r1803529082
